week_1/String/find_all_anagrams/find_all_anagrams.js

```diff
@@ -0,0 +1,67 @@
+// Given two strings long and short, return an array of all the start indices of short's anagrams in long. You may return the answer in any order.
+/**
+ * @param {string} long A longer string
+ * @param {string} short A shorter string
+ * @return {number[]}
+ */
+
+const findAllAnagrams = (long, short) => {
+	if (long.length < short.length) return [];
+
+	const indices = [];
+	// Create a Map that will contain all the values of the shorter string
+	const shortMap = new Map();
+	for (char of short) {
+		shortMap.set(char, shortMap.get(char) + 1 || 1);
+	}
+	console.log(shortMap);
+
+	// Build up a map of equal length from the beginning of the longer string
+	const longMap = new Map();
+	for (let i = 0; i < short.length; i++) {
+		let char = long[i];
+		longMap.set(char, longMap.get(char) + 1 || 1);
+	}
+
+	// Then, if they're the same, add 0 to the array of indices.
+	// To me, it's more intuitive to do this in reverse, because if one value is unequal, it doesn't work,
+	// and it's easier to remove on the conditional in a loop.
+	let startingIndex = 0;
+	indices.push(startingIndex);
+	for (let [key, value] of shortMap) {
+		if (longMap.get(key) !== value) {
+			indices.pop();
```

Comment:
This leaves us with a lot of unnecessary loop iterations. I think a `while` loop would be more appropriate here.

```js
let allMatch = true;
while (allMatch) {
  ... let to false if something doesn't match
 ... break if we're at the end of short
}

if (allAll) {
  indicies.push(startingIndex);
}
```

Reply:
It's also nice to lean into using JavaScript's array methods to be semantic. Using "find" here might work, too.